crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -122,6 +128,171 @@ impl<N: ProviderNodeTypes> BlockchainProvider2<N> {
         (start, end)
     }
 
+    /// Return the last N blocks of state, recreating the [`ExecutionOutcome`].
+    ///
+    /// 1. Iterate over the [`BlockBodyIndices`][reth_db::BlockBodyIndices] table to get all the
+    ///    transaction ids.
+    /// 2. Iterate over the [`StorageChangeSets`][reth_db::StorageChangeSets] table and the
+    ///    [`AccountChangeSets`][reth_db::AccountChangeSets] tables in reverse order to reconstruct
+    ///    the changesets.
+    ///    - In order to have both the old and new values in the changesets, we also access the
+    ///      plain state tables.
+    /// 3. While iterating over the changeset tables, if we encounter a new account or storage slot,
+    ///    we:
+    ///     1. Take the old value from the changeset
+    ///     2. Take the new value from the plain state
+    ///     3. Save the old value to the local state
+    /// 4. While iterating over the changeset tables, if we encounter an account/storage slot we
+    ///    have seen before we:
+    ///     1. Take the old value from the changeset
+    ///     2. Take the new value from the local state
+    ///     3. Set the local state to the value in the changeset
+    ///
+    /// If the range is empty, or there are no blocks for the given range, then this returns `None`.
+    pub fn get_state(
```

Comment:
akin to checks in historical provider this method should check lowest available history blocks in case of a pruned node

Reply:
Now done in `ChangeSetReader` and `StorageChangeSetReader`, ensuring we don't create a db provider without checking prune checkpoints first: https://github.com/paradigmxyz/reth/pull/11265/commits/d7e6542327fe3098cb1e634706ed752fcf2e57eb

Reply:
@Rjected a valid point from @joshieDo here. in scope of this function we perform a ton of lookups by block number, but there is no consistency guarantee here